lib/write-set-tokens.ts

```diff
@@ -0,0 +1,98 @@
+import { types, Set, Range, Char, SetTokens } from './types';
+import * as sets from './sets-lookup';
+
+/**
+ * Takes character code and returns character to be displayed in a set
+ * @param {number} charCode Character code of set element
+ * @param {boolean} firstChar True if character is the first character of the set
+ * @param {boolean} lastChar True if character is the last character of the set
+ * @returns {string} The string for the sets character
+ */
+export function setChar(charCode: number, firstChar: boolean, lastChar = false): string {
+  // We only need to negate first internal '^' if token.not is false
+  return firstChar && charCode === 94 ? '\\^' :
+    charCode === 92 ? '\\\\' :
+      charCode === 93 ? '\\]' :
+        charCode === 45 && !lastChar ? '\\-' :
+          String.fromCharCode(charCode);
+}
+
+/**
+ * Test if a character set matches a 'set-lookup'
+ * @param {SetTokens} set The set to be tested
+ * @param {Record<string | number, boolean>} map The predefined 'set-lookup'
+ * @returns {boolean} True if the character set corresponds to the 'set-lookup'
+ */
+function sameSet(set: SetTokens, map: Record<string | number, boolean>): boolean {
+  for (const elem of set) {
+    if (elem.type === types.SET) {
+      return false;
+    }
+    const key = elem.type === types.CHAR ? elem.value : `${elem.from}-${elem.to}`;
+    if (map[key]) {
+      map[key] = false;
+    } else {
+      return false;
+    }
+  }
+  return true;
+}
+
+/**
+ * Writes the tokens for a set
+ * @param {Set} set The set to display
+ * @param {boolean} firstChar Whether subset contains the first character of the set
+ * @param {boolean} nested Whether the token is nested inside another set token
```

Comment:
would prefer boolean variables to be named `isFirstChar`, `isNested`, `isLastChar`, and `isSameSet`. to make it clearer what they represent while looking through code.